src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs

```diff
@@ -34,207 +45,165 @@ internal sealed class SlidingTextWindow : IDisposable
+        /// access time to characters).
+        /// </summary>
+        public SourceText Text { get; }
 
-        // Example for the above variables:
-        // The text starts at 0.
-        // The window onto the text starts at basis.
-        // The current character is at (basis + offset), AKA the current "Position".
-        // The current lexeme started at (basis + lexemeStart), which is <= (basis + offset)
-        // The current lexeme is the characters between the lexemeStart and the offset.
+        /// <summary>
+        /// Absolute end position of <see cref="Text"/>.  Attempts to read at or past this position will 
+        /// produce <see cref="InvalidCharacter"/>.
+        /// </summary>
+        private readonly int _textEnd;
 
-        private readonly StringTable _strings;
+        /// <summary>
+        /// The current position in <see cref="Text"/> that we are reading characters from.  This is the absolute position in the source text.
+        /// </summary>
+        private int _positionInText;
 
-        private static readonly ObjectPool<char[]> s_windowPool = new ObjectPool<char[]>(() => new char[DefaultWindowLength]);
+        /// <summary>
+        /// Current segment of readable characters.  The backing store for this is an array given out by <see cref="s_windowPool"/>.
+        /// The length of this will normally be <see cref="DefaultWindowLength"/>, but may be smaller if we are at the end of the file
+        /// and there are not enough characters left to fill the window.
+        /// </summary>
+        private ArraySegment<char> _characterWindow;
+
+        /// <summary>
+        /// Where the current character window starts in the source text.  This is the absolute position in the source text.
+        /// In other words, if this is 2048, then that means it represents the chunk of characters starting at position 2048
+        /// in the source text. <c>CharacterWindow.Count</c> represents how large the chunk is.  Characters
+        /// <c>[0, CharacterWindow.Count)</c> are valid characters within the window, and represents
+        /// the chunk <c>[CharacterWindowStartPositionInText, CharacterWindowEndPositionInText)</c> in <see cref="Text"/>.
+        /// </summary>
+        private int _characterWindowStartPositionInText;
+
+        private bool _disposed;
+
+        private readonly StringTable _strings;
 
         public SlidingTextWindow(SourceText text)
         {
-            _text = text;
-            _basis = 0;
-            _offset = 0;
+            this.Text = text;
             _textEnd = text.Length;
             _strings = StringTable.GetInstance();
-            _characterWindow = s_windowPool.Allocate();
-            _lexemeStart = 0;
+            _characterWindow = new(s_windowPool.Allocate());
+
+            // Read the first chunk of the file into the character window.
+            this.ReadChunkAt(0);
```

Comment:
Something i noticed from before.  first token lexing would always fail to quick-lex, since there was no actual chars in the buffer.   this just simplifies things as the charwindow is now only 0 length if we're actually at the end of the text.